### PR TITLE
[WIP] Forward declare Tensor and similar classes in one place.

### PR DESCRIPTION
**Pull request by drwells**
I noticed, while working on #3805 and #3811, that `symmetric_tensor.h` and `tensor.h` are included in so many files that (nearly) the entire library must be recompiled if one of those files changes. I did a bit of work to try and make the situation better by adding more forward declarations.

Since `DerivativeForm`, `Point`, `Tensor`, and `SymmetricTensor` all have default template arguments, they cannot be forward declared (properly) more than once in a translation unit so I put the forward declarations with optional arguments in a separate file, `tensor_classes_fwd.h` (I picked the name to look like `iosfwd.h`). I would be happy to hear about a better approach or name for this.

Edit: fixed typo

### CODE REVIEW

File: include/deal.II/base/derivative_form.h [hunk 1]
```diff
@@ -52,7 +53,7 @@ DEAL_II_NAMESPACE_OPEN
  *
  * @author Sebastian Pauletti, 2011, Luca Heltai, 2015
  */
-template <int order, int dim, int spacedim, typename Number=double>
+template <int order, int dim, int spacedim, typename Number>
```
Comment (bangerth) [inline include/deal.II/base/derivative_form.h]: I don't have time this evening to look through the whole rest, but have you checked what doxygen makes of this? Does it still show the default argument?

**Comment by jppelteret** [comment]
Thanks for doing this @drwells! I think its a great idea. I was scratching around the tensor classes a couple of weeks ago and testing the changes was quite a time consuming exercise :-)

**Comment by drwells** [comment]
@bangerth Doxygen no longer shows the default argument with this approach. I'll poke around at this a little more to see if I can do the forward declarations and make Doxygen and the IDEs know about the default argument.

**Comment by drwells** [comment]
Amusingly, if you look at https://www.dealii.org/developer/doxygen/deal.II/classTensor.html it says that `Tensor` is defined in `mpi.h` :-)

**Comment by drwells** [comment]
This PR massively confuses at least one IDE (kdevelop), so for the sake of not driving our tooling crazy I think that I am going to close this.

There are some related doxygen fixes (i.e., we cannot currently see the default template arguments in a few places) that I will post in a smaller PR later.

**Pull request closed without merging**